### ISSUE TITLE
CHP input sliders break when steam output conversion is 0

### ISSUE THREAD

**Issue by mabijkerk** [quintel/etsource#2716]
**Background**
The inputs and outputs of CHPs in the ETM for the EU datasets are sourced from the Eurostat energy balance. Using the fuel inputs and the electricity and heat production, the electricity and heat conversions (efficiencies) are calculated.

The energy balance only registers **sold heat**. Heat that is produced and consumed on site is not registered, but instead, the fuel required to produce that heat is moved to the final demand. This means that for some CHPs, typically for autoproducers, the heat production that is shown on the energy balance may be 0, even though in reality these CHPs are likely to have produced heat. Notifying @AlexanderWirtz who can add to this explanation if necessary.

See this example of the 2019 energy balance for Belgium:
<img width="651" alt="Screenshot 2022-07-22 at 15 05 37" src="https://user-images.githubusercontent.com/67338510/180445253-2625619a-8dd3-4689-9938-859cc0099460.png">

**The issue**
The input sliders of CHPs are set using the `steam_hot_water_output_conversion`. If the heat production is 0, for example for the `industry_chp_ultra_supercritical_coal` in Belgium, then the output conversion will also be 0. See the [input slider](https://github.com/quintel/etsource/blob/master/inputs/supply/heat/industry/capacity_of_industry_chp_ultra_supercritical_coal.ad) for this coal CHP below:

```ruby
- query =
    demand_per_mw = V(industry_chp_ultra_supercritical_coal,"full_load_hours * MJ_PER_MWH / steam_hot_water_output_conversion");
    EACH(
        UPDATE(
            V(industry_chp_ultra_supercritical_coal),
            number_of_units,
            USER_INPUT() / V(industry_chp_ultra_supercritical_coal,heat_output_capacity)
        ),
        UPDATE(
            L(industry_chp_ultra_supercritical_coal),
            preset_demand,
            USER_INPUT() * demand_per_mw
    ))
```

When input sliders for CHP capacities are set for countries for which the heat efficiency (steam_hot_water_output_conversion) of those CHP's are 0, the model breaks.

**Solution**
The root cause of this issue is that, in contrast to the nl2019 dataset, we do not correct for unsold heat in the EU datasets. While it is good to have this root issue in scope, fixing it would be more of a long-term improvement as it is quite complex. @AlexanderWirtz would you agree?

For one of our clients, fixing this issue is rather urgent, meaning that a quicker solution is required. I therefore propose to update the input sliders in such a way that they can handle a steam_hot_water_output_conversion of 0. A possible route to tackle this is to create an IF statement in the input slider, that sets the `demand_per_mw` using the `electricity_output_conversion` if the `steam_hot_water_output_conversion` is 0. **Note**: according to the front-end the input slider sets the heat output capacity, so I'm not sure if switching to the electricity_output_conversion is entirely 'correct', but in this case I don't see an alternative that is equally viable.

**Comment by marliekeverweij** [comment]
I had a look at the issue and I noticed something more. For Belgium I do not see any industrial CHP capacity installed in present, even though these CHPs are running (see screenshot of Dataset Manager) and the other CHPs (except for the coal CHP) have a `steam_hot_water_output_conversion` > 0.0. Do you know what going on here @antw ?

![Screenshot 2022-07-25 at 17 02 32](https://user-images.githubusercontent.com/32833996/180809580-7af76080-08d9-4f77-b20d-3ceea94ce3eb.png)
![Screenshot 2022-07-25 at 17 03 54](https://user-images.githubusercontent.com/32833996/180810185-c0fd07f7-cff6-4edf-bd6b-1a997d001b15.png)

**Comment by antw** [comment]
> For Belgium I do not see any industrial CHP capacity installed in present, even though these CHPs are running

Does the graph_values.yml for the region have the expected values from ETLocal?

**Comment by marliekeverweij** [comment]
>Does the graph_values.yml for the region have the expected values from ETLocal?

Yes, you can find the demand, output efficiencies and flh of the gas turbine CHP [here](https://github.com/quintel/etsource/blob/master/datasets/BE_belgium/graph_values.yml#L540-L545)

**Comment by marliekeverweij** [comment]
Okay, never mind, now the capacities for Belgium do show up!

![Screenshot 2022-07-26 at 12 06 44](https://user-images.githubusercontent.com/32833996/180981077-2685433a-8b29-4ae2-aa29-e0538bf45b0f.png)

**Comment by marliekeverweij** [comment]
I did some research for which countries (and how many CHPs) this issue occurs (by searching for
```
output:
    :steam_hot_water: 0.0
```
in ETSource

Countries: 

- Belgium - 1
- Greece - 2
- Spain - 10
- Croatia - 1
- Hungary - 2
- Ireland - 6
- Portugal - 3
- Slovakia - 1
- UK - 4

Which CHPs? **Almost all: 12 of 15 CHPs in ETM**

- energy_chp_local_wood_pellets
- energy_chp_supercritical_waste_mix
- energy_chp_ultra_supercritical_coal
- energy_chp_combined_cycle_network_gas
- energy_chp_local_engine_network_gas
- energy_chp_ultra_supercritical_lignite
- industry_chp_ultra_supercritical_coal
- industry_chp_wood_pellets
- industry_chp_turbine_gas_power_fuelmix
- industry_chp_combined_cycle_gas_power_fuelmix
- industry_chp_ultra_supercritical_coal
- industry_chp_engine_gas_power_fuelmix

I think only the input statements of the industrial CHPs are the issue, because these sliders set the heat output capacity instead of the electricity output capacity. The `heat_output_capacity` and `steam_hot_water_output_conversion` are both `0.0` for the CHPs in the countries above and this triggers the error.

The solution that @mabijkerk proposes above will not work (just tested it), because there is no `heat_output_capacity`, so this also has to be replaced with the `electricity_output_capacity`. This actually means that these sliders will set the **electricity output capacity** instead of the **heat output capacity**. This is a good solution! However, this impacts all scenarios that set these sliders. We can write a scenario migration for this, but this is not a straight-forward one.

Another solution we can think of is to give the datasets with these errors a default `steam_hot_water_output_conversion` that is higher than zero. We can add this to the IF() statements in ETLocal, that already give a default efficiency to CHPs that are not running for that country. However, in this case we should be careful that the total `output_conversrion` will not be higher than `1.0`. And this changes the datasets for these countries, because these CHPs will now also produce heat!

So it's not clear yet which way to go! Would like to have a meeting with @AlexanderWirtz tomorrow to discuss this.

**Comment by marliekeverweij** [comment]
After today's meeting with @AlexanderWirtz we would like to propose to:
1. Adjust the 5 industry CHP sliders: setting the **electricity output capacity** instead of the **heat output capacity** (including a scenario migration)
2. Add heat efficiency sliders for these 5 industry CHPs, so users can set a `steam_hot_water_output_conversion` > `0.0`

These two improvements will make sure that the sliders will function again ánd makes it possible to set realistic heat efficiency values for these CHPs.

> Another solution we can think of is to give the datasets with these errors a default steam_hot_water_output_conversion that is higher than zero. We can add this to the IF() statements in ETLocal, that already give a default efficiency to CHPs that are not running for that country. However, in this case we should be careful that the total output_conversrion will not be higher than 1.0. And this changes the datasets for these countries, because these CHPs will now also produce heat!

This solution would create a sort of double-counting of energy: we would be adding more heat to the energy system, but the final demand will not be lower. We thought about only increasing the heat efficiency with a small percentage and lowering the final demand (due to unsold heat) with a small percentage, but this is not easy to implement in an efficient and structured way for all EU datasets.
